docs/metrics.md

```diff
@@ -622,3 +628,23 @@ class CustomHandlerExample:
         # except this time with gauge metric type object
         metrics.add_size("GaugeModelMetricNameExample", 42.5)
 ```
+
+## Backwards compatibility warnings
+1. Starting [v0.6.1](https://github.com/pytorch/serve/releases/tag/v0.6.1), the `add_metric` API signature changed\
+   from [add_metric(name, value, unit, idx=None, dimensions=None)](https://github.com/pytorch/serve/blob/61f1c4182e6e864c9ef1af99439854af3409d325/ts/metrics/metrics_store.py#L184)\
```

Comment:
so the old metrics were always counters? If that's the case then keeping BC automatically shouldn't be too hard

Reply:
The [prior metrics implementation](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/metric.py#L16-L29) did not have `types` associated with them. The [new metrics implementation](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/caching_metric.py#L24) does add support for [MetricTypes](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/metric_type_enum.py#L7).

In addition, while the prior metrics implementation did not have a way to specify metrics and their specifications (name, unit, dimension names and type) in a central configuration file, the new metrics implementation introduced this, as a result which, the semantics of `add_metric` method was changed
**from**: [Create a metric object and store in a list to emit](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/metrics_store.py#L184)
**to**: [Add a metric object consisting only of its specifications (name, unit, dimension names and type) to a metics cache](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/metric_cache_yaml_impl.py#L97-L103). The dimension values are provided at the time of updating a metric using the [add_or_update](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/caching_metric.py#L107-L112) method.

Reply:
Couple of options to ensure backwards compatibility are as follows:
1) Introduce a new method in [metric_cache_abstract.py](https://github.com/pytorch/serve/blob/96450b9d0ab2a7290221f0e07aea5fda8a83efaf/ts/metrics/metric_cache_abstract.py), say `add_metric_bc` which has the same signature as that of the old [add_metric](https://github.com/pytorch/serve/blob/61f1c4182e6e864c9ef1af99439854af3409d325/ts/metrics/metrics_store.py#L184) API. This method can internally call `add_metric` and then `add_or_update` on the metric object. The default metric type in this case would be `counter`.
2) Change the name of the new `add_metric` method to `add_metric_to_cache` and reimplement the `add_metric` method to have the same signature as the old implementation. Then, the `add_metric` API can internally call `add_metric_to_cache` method and the `add_or_update` on the metric object.

Please share your thoughts on these approaches.

Reply:
I like 2, `add_metric_to_cache()` seems more like an internal detail whereas what a user wants to do is `add_metric()`.

Reply:
Draft PR to implement backwards compatibility for `add_metric` API: https://github.com/pytorch/serve/pull/2525